File: java/1140. Stone Game II.java
class Solution {
    public int stoneGameII(int[] piles) {
        int[] suffix = Arrays.copyOf(piles, piles.length);
        for (int i = piles.length-2; i >= 0; i--) {
            suffix[i] += suffix[i+1];
        }
        Integer[][] cache = new Integer[piles.length][piles.length];
        return helper(piles, 0, 1, suffix, cache);
    }
    private int helper(int[] piles, int index, int m, int[] suffix, Integer[][] cache) {
        if (index + 2*m >= piles.length) return suffix[index];
        if (cache[index][m] != null) return cache[index][m];
        int answer = Integer.MAX_VALUE;
        for (int i = 1; i <= 2*m; i++) {
            answer = Math.min(answer, helper(piles, index+i, Math.max(i, m), suffix, cache));
        }
        return cache[index][m] = suffix[index] - answer;
    }
}

File: go/1140. Stone Game II.go
func stoneGameII(piles []int) int {
	suffix := make([]int, len(piles))
	copy(suffix, piles)
	for i := len(piles)-2; i >= 0; i-- {
		suffix[i] += suffix[i+1]
	}
	cache := make([][]int, len(piles))
	for i := range cache {
		cache[i] = make([]int, len(piles))
		for j := range cache[i] {
			cache[i][j] = -1
		}
	}
	return helper(piles, 0, 1, suffix, cache)
}

func helper(piles []int, index, m int, suffix []int, cache [][]int) int {
	if index + 2*m >= len(piles) {
		return suffix[index]
	}
	if cache[index][m] != -1 { return cache[index][m] }
	answer := math.MaxInt32
	for i := 1; i <= 2*m; i++ {
		answer = min(answer, helper(piles, index+i, max(i, m), suffix, cache))
	}
	cache[index][m] = suffix[index] - answer
	return cache[index][m]
}